File: src/main/java/io/spring/springbatch/StepBuilderConfiguration.java
package io.spring.springbatch;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.batch.core.Job;
import org.springframework.batch.core.Step;
import org.springframework.batch.core.configuration.annotation.JobBuilderFactory;
import org.springframework.batch.core.configuration.annotation.StepBuilderFactory;
import org.springframework.batch.core.job.builder.FlowBuilder;
import org.springframework.batch.core.job.flow.Flow;
import org.springframework.batch.core.launch.support.RunIdIncrementer;
import org.springframework.batch.item.ItemProcessor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import static org.springframework.batch.repeat.RepeatStatus.FINISHED;

@Slf4j
@Configuration
@RequiredArgsConstructor
public class StepBuilderConfiguration {
    private final JobBuilderFactory jobBuilderFactory;
    private final StepBuilderFactory stepBuilderFactory;

    @Bean
    public Job batchJob() {
        return this.jobBuilderFactory.get("batchJob")
                .incrementer(new RunIdIncrementer())
                .start(this.step1())
                .next(this.step2())
                .next(this.step3())
                .build();
    }

    @Bean
    public Step step1() {
        return this.stepBuilderFactory.get("step1")
                .tasklet((contribution, chunkContext) -> {
                    log.info("step1 has executed");
                    return FINISHED;
                })
                .build();
    }

    @Bean
    public Step step2() {
        return this.stepBuilderFactory.get("step2")
                .<String, String>chunk(3)
                .reader(() -> null)
                .processor((ItemProcessor<String, String>) item -> null)
                .writer(items -> {

                })
                .build();
    }

    @Bean
    public Step step3() {
        return this.stepBuilderFactory.get("step3")
                .partitioner(this.step1())
                .gridSize(2)
                .build();
    }

    @Bean
    public Step step4() {
        return this.stepBuilderFactory.get("step4")
                .job(this.job())
                .build();
    }


    @Bean
    public Step step5() {
        return this.stepBuilderFactory.get("step5")
                .flow(this.flow())
                .build();
    }

    @Bean
    public Job job() {
        return this.jobBuilderFactory.get("job")
                .incrementer(new RunIdIncrementer())
                .start(this.step1())
                .next(this.step2())
                .next(this.step3())
                .build();
    }

    public Flow flow() {
        final FlowBuilder<Flow> flowBuilder = new FlowBuilder<>("flow");
        flowBuilder.start(this.step2()).end();
        return flowBuilder.build();
    }

}

File: src/main/kotlin/com/spring/batch/StepBuilderConfiguration.kt
package com.spring.batch

import io.klogging.NoCoLogging
import org.springframework.batch.core.Job
import org.springframework.batch.core.Step
import org.springframework.batch.core.job.builder.FlowBuilder
import org.springframework.batch.core.job.builder.JobBuilder
import org.springframework.batch.core.job.flow.Flow
import org.springframework.batch.core.launch.support.RunIdIncrementer
import org.springframework.batch.core.repository.JobRepository
import org.springframework.batch.core.step.builder.StepBuilder
import org.springframework.batch.repeat.RepeatStatus
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.transaction.PlatformTransactionManager

@Configuration
class StepBuilderConfiguration(
    private val jobRepository: JobRepository,
    private val platformTransactionManager: PlatformTransactionManager,
) : NoCoLogging {
    @Bean
    fun batchJob(
        step1: Step,
        step2: Step,
        step3: Step,
    ): Job = JobBuilder("batchJob", this.jobRepository)
        .start(step1)
        .next(step2)
        .next(step3)
        .incrementer(RunIdIncrementer())
        .build()

    @Bean
    fun job(
        step1: Step,
        step2: Step,
        step3: Step,
    ): Job = JobBuilder("job", this.jobRepository)
        .start(step1)
        .next(step2)
        .next(step3)
        .build()

    @Bean
    fun step1(
    ): Step = StepBuilder("step1", this.jobRepository).tasklet({ _, _ ->
        logger.info { "Hello, World!" }
        RepeatStatus.FINISHED
    }, this.platformTransactionManager).build()

    @Bean
    fun step2(
    ): Step = StepBuilder("step2", this.jobRepository)
        .chunk<String, String>(10, platformTransactionManager)
        .reader { null }
        .processor { null }
        .writer {}
        .build()

    @Bean
    fun step3(
        step1: Step
    ): Step = StepBuilder("step3", this.jobRepository)
        .partitioner(step1)
        .gridSize(2)
        .build()

    @Bean
    fun step4(
        job: Job
    ): Step = StepBuilder("step4", this.jobRepository)
        .job(job)
        .build()

    @Bean
    fun step5(
        flow: Flow,
    ): Step = StepBuilder("step4", this.jobRepository)
        .flow(flow)
        .build()

    @Bean
    fun flow(
        step2: Step
    ): Flow {
        val flowBuilder = FlowBuilder<Flow>("flow")
        flowBuilder.start(step2).end()
        return flowBuilder.build()
    }

}